### PR TITLE
CORE-7101 Current sandbox thread local

### PR DESCRIPTION
**Pull request by lankydan**
Create `CurrentSandboxGroupContext` to set, get and remove the current `SandboxGroupContext` on a thread.

This service can then be OSGI injected into other services to retrieve the current sandbox. The thread local concept is private to the service.

### CODE REVIEW

File: components/virtual-node/sandbox-group-context-service/src/test/kotlin/net/corda/sandboxgroupcontext/impl/SandboxGroupServiceImplTest.kt [hunk 1]
```diff
@@ -39,7 +39,7 @@ class StubSandboxGroupContextCache: SandboxGroupContextCache {
     }
 }
 
-class SandboxGroupContextServiceImplTest {
+class SandboxGroupServiceImplTest {
```
Comment (chrisr3) [inline components/virtual-node/sandbox-group-context-service/src/test/kotlin/net/corda/sandboxgroupcontext/impl/SandboxGroupServiceImplTest.kt]: Is this renaming a remnant from your earlier PR?

File: components/virtual-node/sandbox-group-context-service/src/main/kotlin/net/corda/sandboxgroupcontext/service/impl/CurrentSandboxGroupContextImpl.kt [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+            // local is being incorrectly set or removed.
+            log.warn(
+                "No current sandbox group context to remove for thread [${Thread.currentThread().name}]",
+                IllegalStateException("No current sandbox group context to remove for thread")
+            )
+        }
+    }
+}
+
+private val currentSandboxGroupContext = ThreadLocal<SandboxGroupContext>()
```
Comment (chrisr3) [inline components/virtual-node/sandbox-group-context-service/src/main/kotlin/net/corda/sandboxgroupcontext/service/impl/CurrentSandboxGroupContextImpl.kt]: Should this field belong to the `companion object`?

File: components/virtual-node/ledger-persistence-service-impl/src/main/kotlin/net/corda/ledger/persistence/impl/internal/ConsensualLedgerMessageProcessor.kt [hunk 3]
```diff
@@ -25,6 +25,7 @@ import net.corda.v5.base.util.debug
 import net.corda.v5.cipher.suite.DigestService
 import net.corda.v5.cipher.suite.merkle.MerkleTreeProvider
 import net.corda.virtualnode.toCorda
+import java.nio.ByteBuffer
```
Comment (chrisr3) [inline components/virtual-node/ledger-persistence-service-impl/src/main/kotlin/net/corda/ledger/persistence/impl/internal/ConsensualLedgerMessageProcessor.kt]: Your import statements appear to be playing "musical chairs"... :dizzy_face:

**Review by chrisr3 — DISMISSED** [review]
I'm unclear on some of the finer details, but there does appear to be some lingering residue from an earlier PR in here.

File: components/virtual-node/ledger-persistence-service-impl/src/main/kotlin/net/corda/ledger/persistence/impl/internal/ConsensualLedgerMessageProcessor.kt (hunk 3 above)
Comment (lankydan) [reply components/virtual-node/ledger-persistence-service-impl/src/main/kotlin/net/corda/ledger/persistence/impl/internal/ConsensualLedgerMessageProcessor.kt]: Yeah, I screwed up my formatting config a while back and finally sorted it out. So this is fixing previous errors.

File: components/virtual-node/sandbox-group-context-service/src/test/kotlin/net/corda/sandboxgroupcontext/impl/SandboxGroupServiceImplTest.kt (hunk 1 above)
Comment (lankydan) [reply components/virtual-node/sandbox-group-context-service/src/test/kotlin/net/corda/sandboxgroupcontext/impl/SandboxGroupServiceImplTest.kt]: Yes, that should have disappeared now.

File: components/virtual-node/sandbox-group-context-service/src/main/kotlin/net/corda/sandboxgroupcontext/service/impl/CurrentSandboxGroupContextImpl.kt (hunk 2 above)
Comment (lankydan) [reply components/virtual-node/sandbox-group-context-service/src/main/kotlin/net/corda/sandboxgroupcontext/service/impl/CurrentSandboxGroupContextImpl.kt]: Don't think it matters in this case, but I can move it in.

File: components/virtual-node/sandbox-group-context-service/src/main/kotlin/net/corda/sandboxgroupcontext/service/impl/CurrentSandboxGroupContextImpl.kt [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+import net.corda.v5.serialization.SingletonSerializeAsToken
+import org.osgi.service.component.annotations.Component
+
+@Component(service = [CurrentSandboxGroupContext::class, SingletonSerializeAsToken::class])
+class CurrentSandboxGroupContextImpl : CurrentSandboxGroupContext, SingletonSerializeAsToken {
+
+    private companion object {
+        val log = contextLogger()
+
+        val currentSandboxGroupContext = ThreadLocal<SandboxGroupContext>()
```
Comment (chrisr3) [inline components/virtual-node/sandbox-group-context-service/src/main/kotlin/net/corda/sandboxgroupcontext/service/impl/CurrentSandboxGroupContextImpl.kt]: Do these need to be public? They look like things that should be `private`.

File: components/virtual-node/sandbox-group-context-service/src/main/kotlin/net/corda/sandboxgroupcontext/service/impl/CurrentSandboxGroupContextImpl.kt (hunk 2 above)
Comment (chrisr3) [reply components/virtual-node/sandbox-group-context-service/src/main/kotlin/net/corda/sandboxgroupcontext/service/impl/CurrentSandboxGroupContextImpl.kt]: Anything declared outside the file becomes part of `CurrentSandboxGroupContextImplKt.class`, and an extra class seem like overkill when we already have a `companion object`.

File: components/virtual-node/sandbox-group-context-service/src/main/kotlin/net/corda/sandboxgroupcontext/service/impl/CurrentSandboxGroupContextImpl.kt (hunk 4 above)
Comment (lankydan) [reply components/virtual-node/sandbox-group-context-service/src/main/kotlin/net/corda/sandboxgroupcontext/service/impl/CurrentSandboxGroupContextImpl.kt]: The whole companion object is private.

**Review by blsemo — DISMISSED** [review]
Looks good to me

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/fiber/FlowFiberImpl.kt [hunk 5]
```diff
@@ -52,6 +53,7 @@ class FlowFiberImpl(
 
     @Suspendable
     override fun run() {
+        setCurrentSandboxGroupContext()
```
Comment (mnesbit) [inline components/flow/flow-service/src/main/kotlin/net/corda/flow/fiber/FlowFiberImpl.kt]: need a remove at the end (possibly in an outer try finally?) to handle when we exit the fiber exceptionally

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/fiber/FlowFiberImpl.kt [hunk 6]
```diff
@@ -156,6 +160,7 @@ class FlowFiberImpl(
         if (flowStackItem.sessionIds.isNotEmpty()) {
             suspend(FlowIORequest.SubFlowFinished(flowStackItem.sessionIds.toList()))
         }
+        removeCurrentSandboxGroupContext()
```
Comment (mnesbit) [inline components/flow/flow-service/src/main/kotlin/net/corda/flow/fiber/FlowFiberImpl.kt]: Not sure this is correct. Wouldn't mind the logic explaining. We are still firmly inside the sandbox here.

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/fiber/FlowFiberImpl.kt [hunk 7]
```diff
@@ -169,6 +174,7 @@ class FlowFiberImpl(
         if (flowStackItem.sessionIds.isNotEmpty()) {
             suspend(FlowIORequest.SubFlowFailed(throwable, flowStackItem.sessionIds.toList()))
         }
+        removeCurrentSandboxGroupContext()
```
Comment (mnesbit) [inline components/flow/flow-service/src/main/kotlin/net/corda/flow/fiber/FlowFiberImpl.kt]: same as above I think this is removing the sandbox context a bit early

File: components/virtual-node/sandbox-group-context-service/src/main/kotlin/net/corda/sandboxgroupcontext/service/impl/CurrentSandboxGroupContextImpl.kt [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+            current
+        } else {
+            log.error("No current sandbox group context set for thread [${Thread.currentThread().name}]")
+            throw IllegalStateException("No current sandbox group context set for thread")
+        }
+    }
+
+    override fun remove() {
+        val current = currentSandboxGroupContext.get()
+        currentSandboxGroupContext.remove()
```
Comment (mnesbit) [inline components/virtual-node/sandbox-group-context-service/src/main/kotlin/net/corda/sandboxgroupcontext/service/impl/CurrentSandboxGroupContextImpl.kt]: I vaguely remember remove being quite expensive. It may be better to set to null (which presumably requires a nullable type in definition)

Comment (lankydan) [reply components/virtual-node/sandbox-group-context-service/src/main/kotlin/net/corda/sandboxgroupcontext/service/impl/CurrentSandboxGroupContextImpl.kt]: Changed.

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/fiber/FlowFiberImpl.kt (hunk 6 above)
Comment (lankydan) [reply components/flow/flow-service/src/main/kotlin/net/corda/flow/fiber/FlowFiberImpl.kt]: Moved.

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/fiber/FlowFiberImpl.kt (hunk 5 above)
Comment (lankydan) [reply components/flow/flow-service/src/main/kotlin/net/corda/flow/fiber/FlowFiberImpl.kt]: Done.

**Review by chrisr3 — COMMENTED** [review]
Can this wait until after #2346 is merged please? This will make the `ConsensualLedgerMessageProcessor` integration tests' behaviour more realistic in that they will test the _actual_ `EntityManagerFactory` created by `EntitySandboxService`. (And with a genuine CPB too!)

File: components/virtual-node/sandbox-group-context-service/src/main/kotlin/net/corda/sandboxgroupcontext/service/impl/CurrentSandboxGroupContextImpl.kt [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+import net.corda.v5.serialization.SingletonSerializeAsToken
+import org.osgi.service.component.annotations.Component
+
+@Component(service = [CurrentSandboxGroupContext::class, SingletonSerializeAsToken::class])
+class CurrentSandboxGroupContextImpl : CurrentSandboxGroupContext, SingletonSerializeAsToken {
+
+    private companion object {
+        val log = contextLogger()
+
+        val currentSandboxGroupContext = ThreadLocal<SandboxGroupContext?>()
```
Comment (chrisr3) [inline components/virtual-node/sandbox-group-context-service/src/main/kotlin/net/corda/sandboxgroupcontext/service/impl/CurrentSandboxGroupContextImpl.kt:18]: Do you want to annotate both of these with `@JvmField`?

File: components/persistence/entity-processor-service-impl/src/main/kotlin/net/corda/entityprocessor/impl/internal/EntityMessageProcessor.kt [hunk 10]
```diff
@@ -68,9 +70,12 @@ class EntityMessageProcessor(
                     try {
                         val holdingIdentity = request.holdingIdentity.toCorda()
                         val sandbox = entitySandboxService.get(holdingIdentity)
+                        currentSandboxGroupContext.set(sandbox)
                         processRequestWithSandbox(sandbox, request)
```
Comment (chrisr3) [inline components/persistence/entity-processor-service-impl/src/main/kotlin/net/corda/entityprocessor/impl/internal/EntityMessageProcessor.kt]: I thought that `CurrentSandboxGroupContext` was only useful for `FLOW` sandboxes, due to Quasar fibers?! :confused:

File: libs/virtual-node/sandbox-group-context/src/main/kotlin/net/corda/sandboxgroupcontext/CurrentSandboxGroupContext.kt [hunk 11]
```diff
@@ -0,0 +1,30 @@
+package net.corda.sandboxgroupcontext
+
+/**
+ * [CurrentSandboxGroupContext] allows the current sandbox's [SandboxGroupContext] to be set, retrieved and removed.
+ *
+ * [CurrentSandboxGroupContext] relies on a [ThreadLocal] to provide access to the current sandbox wherever this service is used.
+ *
+ * Generally, few callers should call [set] and [remove] and should be reserved for the setup and cleanup code that is run when setting up
+ * a new sandbox (or retrieving from one a cache) and cleaning up resources when the thread using the sandbox has
+ * completed processing.
```
Comment (chrisr3) [inline libs/virtual-node/sandbox-group-context/src/main/kotlin/net/corda/sandboxgroupcontext/CurrentSandboxGroupContext.kt]: Can you rephrase this comment please? Its meaning is escaping me :dash:.

File: testing/test-serialization/src/main/kotlin/net/corda/internal/serialization/amqp/helper/TestFlowFiberServiceWithSerialization.kt [hunk 12]
```diff
@@ -9,22 +9,25 @@ import net.corda.internal.serialization.amqp.SerializerFactory
 import net.corda.membership.read.MembershipGroupReader
 import net.corda.sandbox.SandboxGroup
 import net.corda.v5.base.types.MemberX500Name
+import net.corda.sandboxgroupcontext.CurrentSandboxGroupContext
```
Comment (chrisr3) [inline testing/test-serialization/src/main/kotlin/net/corda/internal/serialization/amqp/helper/TestFlowFiberServiceWithSerialization.kt]: This new `import` is incorrectly sorted :smiling_imp:.

**Review by chrisr3 — COMMENTED** [review]
I am slightly confused by a few things, and not least why we need this for Quasar-free `PERSISTENCE` sandboxes.

File: components/persistence/entity-processor-service-impl/src/main/kotlin/net/corda/entityprocessor/impl/internal/EntityMessageProcessor.kt (hunk 10 above)
Comment (lankydan) [reply components/persistence/entity-processor-service-impl/src/main/kotlin/net/corda/entityprocessor/impl/internal/EntityMessageProcessor.kt]: Sure, I can take it out, we might add it back in in the future, but it isn't providing any benefit at the moment.

File: components/virtual-node/sandbox-group-context-service/src/main/kotlin/net/corda/sandboxgroupcontext/service/impl/CurrentSandboxGroupContextImpl.kt (hunk 9 above)
Comment (lankydan) [reply components/virtual-node/sandbox-group-context-service/src/main/kotlin/net/corda/sandboxgroupcontext/service/impl/CurrentSandboxGroupContextImpl.kt:18]: Since they're private I don't think theres much upside to doing so

Comment (chrisr3) [reply components/virtual-node/sandbox-group-context-service/src/main/kotlin/net/corda/sandboxgroupcontext/service/impl/CurrentSandboxGroupContextImpl.kt:18]: A `private` property is automatically implemented as a field, whereas the compiler also generates a getter for a `public` property.

**Review by mnesbit — APPROVED** [review]
OK

File: components/ledger/ledger-common-flow/src/main/kotlin/net/corda/ledger/common/flow/impl/transaction/factory/TransactionMetadataFactoryImpl.kt [hunk 13]
```diff
@@ -40,10 +40,8 @@ class TransactionMetadataFactoryImpl @Activate constructor(
 
     // CORE-7127 Get rid of flowFiberService and access CPK information without fiber when the related solution gets
```
Comment (vlajos) [inline components/ledger/ledger-common-flow/src/main/kotlin/net/corda/ledger/common/flow/impl/transaction/factory/TransactionMetadataFactoryImpl.kt:41]: I think this PR fixes this ticket/comment.

File: components/ledger/ledger-consensual-flow/src/main/kotlin/net/corda/ledger/consensual/flow/impl/transaction/factory/ConsensualSignedTransactionFactoryImpl.kt [hunk 14]
```diff
@@ -109,8 +109,7 @@ class ConsensualSignedTransactionFactoryImpl @Activate constructor(
     ): List<List<ByteArray>> {
 
         // TODO CORE-7101 use CurrentSandboxService when it gets available
```
Comment (vlajos) [inline components/ledger/ledger-consensual-flow/src/main/kotlin/net/corda/ledger/consensual/flow/impl/transaction/factory/ConsensualSignedTransactionFactoryImpl.kt:111]: This looks done as well.

**Review by vlajos — APPROVED** [review]

**Review by ronanbrowne — APPROVED** [review]
lgtm (build)

File: components/ledger/ledger-consensual-flow/src/main/kotlin/net/corda/ledger/consensual/flow/impl/transaction/factory/ConsensualSignedTransactionFactoryImpl.kt (hunk 14 above)
Comment (lankydan) [reply components/ledger/ledger-consensual-flow/src/main/kotlin/net/corda/ledger/consensual/flow/impl/transaction/factory/ConsensualSignedTransactionFactoryImpl.kt:111]: I'll quietly delete these TODOs in a random PR.
